cpp/include/raft/interruptible.hpp

```diff
@@ -0,0 +1,246 @@
+    }
+    return thread_store;
+  }
+
+  /**
+   * @brief Cancel any current or next call to `interruptible::synchronize` performed on the
+   * CPU thread given by the `thread_id`
+   *
+   * Note, this function uses a mutex to safely get a cancellation token that may be shared
+   * among multiple threads. If you plan to use it from a signal handler, consider the non-static
+   * `cancel_no_throw()` instead.
+   *
+   * @param [in] thread_id a CPU thread, in which the work should be interrupted.
+   *
+   * @throw raft::cuda_error if a CUDA error happens during recording the interruption event.
+   */
+  static inline void cancel(std::thread::id thread_id) { get_token(thread_id)->cancel(); }
+
+  /**
+   * @brief Cancel any current or next call to `interruptible::synchronize` performed on the
+   * CPU thread given by this `interruptible` token.
+   *
+   * Note, this function does not involve thread synchronization/locks, so it should be safe
+   * to call from a signal handler.
+   *
+   * @throw raft::cuda_error if a CUDA error happens during recording the interruption event.
+   */
+  inline void cancel() { RAFT_CUDA_TRY(cancel_no_throw()); }
+
+  /**
+   * @brief Cancel any current or next call to `interruptible::synchronize` performed on the
+   * CPU thread given by this `interruptible` token.
+   *
+   * Note, this function does not involve thread synchronization/locks and not throw any
+   * exceptions, so it's safe to call from a signal handler.
+   *
+   * @return CUDA error code redirected from `cudaEventRecord`.
+   */
+  auto cancel_no_throw() noexcept -> cudaError_t
+  {
+    // This method is supposed to be called from another thread;
+    // multiple calls to it just override each other, and that is ok - the cancellation request
+    // will be delivered (at least once).
+    cancelled_ = true;
+    return cudaEventRecord(wait_interrupt_, get_cancellation_stream());
+  }
+
+  // don't allow the token to leave the shared_ptr
+  interruptible(interruptible const&) = delete;
+  interruptible(interruptible&&)      = delete;
+  auto operator=(interruptible const&) -> interruptible& = delete;
+  auto operator=(interruptible&&) -> interruptible& = delete;
+
+ private:
+  /** Global registry of thread-local cancellation stores. */
+  static inline std::unordered_map<std::thread::id, std::weak_ptr<interruptible>> registry_;
+  /** Protect the access to the registry. */
+  static inline std::mutex mutex_;
+  /** The only purpose of this stream is to record the interruption events. */
+  static inline auto get_cancellation_stream() -> cudaStream_t
+  {
+    static std::unique_ptr<cudaStream_t, std::function<void(cudaStream_t*)>> cs{
+      []() {
+        auto* stream = new cudaStream_t;
+        RAFT_CUDA_TRY(cudaStreamCreateWithFlags(stream, cudaStreamNonBlocking));
+        return stream;
+      }(),
+      [](cudaStream_t* stream) {
+        cudaStreamDestroy(*stream);
+        delete stream;
+      }};
+    return *cs;
+  }
```

Comment:
You can avoid the dynamic allocation here by overriding the pointer type in the deleter and exploiting the fact that `cudaStream_t` is a pointer and thus satisfies the NullablePointer concept. 

```suggestion
  struct delete_stream{
     using pointer = cudaStream_t;
     void operator()(cudaStream_t s) const noexcept{
        cudaStreamDestroy(s);
      }
  };
  static inline auto get_cancellation_stream() -> cudaStream_t
  {
    static std::unique_ptr<cudaStream_t, delete_stream> cs{
      []() {
        cudaStream_t stream;
        RAFT_CUDA_TRY(cudaStreamCreateWithFlags(&stream, cudaStreamNonBlocking));
        return stream;
      }()};
    return *cs;
  }
```